collated/main/A0139498J.md

```diff
@@ -0,0 +1,1044 @@
+    
+    public static final String MESSAGE_UNDO_SUCCESS = "Undid delete on tasks! %1$s Tasks restored!";
+    
+    private List<Task> deletedTasks;
+    private List<String> displayDeletedTasks;
+    public final List<Integer> targetIndexes;
+    private boolean viewingDoneList;
+
+    public DeleteCommand(List<Integer> targetIndexes) {
+        assert targetIndexes != null;
+        this.targetIndexes = targetIndexes;
+    }
+
+
+    @Override
+    public CommandResult execute() {
+        assert model != null && targetIndexes != null;
+        
+        displayDeletedTasks = new ArrayList<String>();
+        if (!checkIfRedoAction()) {
+            viewingDoneList = model.isCurrentListDoneList();
+        }
+        Collections.sort(targetIndexes);
+        int adjustmentForRemovedTask = 0;
+        deletedTasks = new ArrayList<Task>();
+        
+        if (viewingDoneList) {
+            return deleteFromDoneList(adjustmentForRemovedTask);
+        }
+        else {
+            return deleteFromUndoneList(adjustmentForRemovedTask);
+        }
+    }
+    
+    private CommandResult deleteFromDoneList(int adjustmentForRemovedTask) {
+        for (int targetIndex: targetIndexes) {
+            UnmodifiableObservableList<ReadOnlyTask> lastShownList;
+            if (viewingDoneList) {
+                lastShownList = model.getFilteredDoneTaskList();
+            } else {
+                lastShownList = model.getFilteredUndoneTaskList();
+            }
+    
+            if (lastShownList.size() < targetIndex - adjustmentForRemovedTask) {
+                continue;
+            }
+    
+            ReadOnlyTask taskToDelete = lastShownList.get(targetIndex - adjustmentForRemovedTask - 1);
+    
+            try {
+                model.deleteDoneTask(taskToDelete);
+                deletedTasks.add((Task) taskToDelete);
+            } catch (TaskNotFoundException pnfe) {
+                assert false : "The target task cannot be missing";
+            }
+            displayDeletedTasks.add(taskToDelete.toString());
+            adjustmentForRemovedTask++;
+        }
+        
+       
+        updateHistory();
+        if (displayDeletedTasks.isEmpty()) {
+            indicateAttemptToExecuteIncorrectCommand();
+            return new CommandResult(Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+        }
+        String toDisplay = displayDeletedTasks.toString().replace("[", "").replace("]", "");
+        return (displayDeletedTasks.size() == 1)? 
+                new CommandResult(String.format(MESSAGE_DELETE_ITEM_SUCCESS, toDisplay)):
+                new CommandResult(String.format(MESSAGE_DELETE_ITEMS_SUCCESS, toDisplay));
+    }
+
+
+    private CommandResult deleteFromUndoneList(int adjustmentForRemovedTask) {
+        for (int targetIndex: targetIndexes) {
+            UnmodifiableObservableList<ReadOnlyTask> lastShownList;
+            if (viewingDoneList) {
```

Comment:
much of the code seems to be duplicated between this and the previous method; you can extract common code as a separate method by itself